### ISSUE TITLE
`param`ness is lost in initializer with param argument

### ISSUE THREAD

**Issue by BryantLam** [chapel-lang/chapel#16308]
### Summary of Problem
This code has an initializer with a `param int` argument that does not preserve the `param`ness of that argument.

```chapel
class C {
  param val: int;
}

record R {
  forwarding var data; // or just `var data`

  proc init(param val: int) {
    this.data = forall j in 1..10 do new owned C(val);

    // Above does not work. This will work:
    // param _val = val;
    // this.data = forall j in 1..10 do new owned C(_val);
  }
}

proc main() {
  param val = 8;
  var r = new R(val);
}
```
```terminal
.code.tio.chpl:8: In initializer:
.code.tio.chpl:9: error: unresolved call 'C.init(int(64))'
.code.tio.chpl:1: note: this candidate did not match: C.init(param val: int(64))
.code.tio.chpl:9: note: because non-param call actual argument #1
.code.tio.chpl:2: note: is passed to param formal 'param val: int(64)'
.code.tio.chpl:15: Initializer instantiated as: init(param val = 8)
```
Affects `chpl version` 1.20 through 1.23 pre-release

**Comment by bradcray** [comment]
Wow, that's an embarrassing one!  Thanks for reporting this Bryant.

I was initially thinking this might be a shadow variable issue, but changing the `forall` to a `for` doesn't fix things and should remove the shadow variable.  Also, Bryant's workaround suggests that shadow variables are not involved.

Removing the loop altogether does cause it to work, though, suggesting that it has something to do with passing the param formal to a nested procedure (though why Bryant's workaround wouldn't give the same behavior, I'm unsure about).

Tagging @vasslitvinov and @mppf to see whether they have any insights here, as I think you've done the most with expression-level control flow and argument passing lately (?).

**Comment by vasslitvinov** [comment]
Should be resolved with #16318.
